### PR TITLE
planner: count rune if charset is neither binary nor ascii

### PR DESCRIPTION
**Pull request by xiongjiwei**
close https://github.com/pingcap/tidb/issues/26820
if the charset is neither `binary` nor `ASCII`, count the rune in the string

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

File: util/ranger/ranger.go [hunk 1]
```diff
@@ -498,26 +498,25 @@ func fixPrefixColRange(ranges []*Range, lengths []int, tp []*types.FieldType) bo
...
 	if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
 		colCharset := tp.Charset
 		colValue := v.GetBytes()
-		isUTF8Charset := colCharset == charset.CharsetUTF8 || colCharset == charset.CharsetUTF8MB4
-		if isUTF8Charset {
-			if length != types.UnspecifiedLength && utf8.RuneCount(colValue) > length {
-				rs := bytes.Runes(colValue)
-				truncateStr := string(rs[:length])
+		if colCharset == charset.CharsetBin || colCharset == charset.CharsetASCII {
+			if length != types.UnspecifiedLength && len(colValue) > length {
```
Comment (zimulala) [inline util/ranger/ranger.go:507]: Could we put `length != types.UnspecifiedLength` to line503?

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -529,11 +528,10 @@ func ReachPrefixLen(v *types.Datum, length int, tp *types.FieldType) bool {
 	if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
 		colCharset := tp.Charset
 		colValue := v.GetBytes()
-		isUTF8Charset := colCharset == charset.CharsetUTF8 || colCharset == charset.CharsetUTF8MB4
-		if isUTF8Charset {
-			return length != types.UnspecifiedLength && utf8.RuneCount(colValue) == length
+		if colCharset == charset.CharsetBin || colCharset == charset.CharsetASCII {
+			return length != types.UnspecifiedLength && len(colValue) == length
```
Comment (zimulala) [inline util/ranger/ranger.go:532]: Ditto.

File: planner/core/exhaust_physical_plans_test.go [hunk 3]
```diff
@@ -201,7 +201,7 @@ func (s *testUnitTestSuit) TestIndexJoinAnalyzeLookUpFilters(c *C) {
 			innerKeys:       []*expression.Column{dsSchema.Columns[1]},
 			pushedDownConds: "a = 1 and c > 'a' and c < 'aaaaaa'",
 			otherConds:      "",
-			ranges:          "[(1 NULL \"a\",1 NULL 0x6161]]",
+			ranges:          "[(1 NULL \"a\",1 NULL \"aa\"]]",
```
Comment (zimulala) [inline planner/core/exhaust_physical_plans_test.go:204]: Why did the range change?

Comment (xiongjiwei) [reply planner/core/exhaust_physical_plans_test.go:204]: it is a bug in the original code. https://github.com/pingcap/tidb/blob/19d3b1f020a4108afbd2351a3da6357ab3e056bb/util/ranger/ranger.go#L517-L520
`SetBytes` will set the Datum type to `KindBytes` which makes the charset to be `binary`. L518 is always false.

Comment (zimulala) [reply planner/core/exhaust_physical_plans_test.go:204]: But the result was changed before this PR dealt with the bug. What is the reason?

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**